---
title: "Differential gene expression and preliminary ORA on GSE185657"
subtitle: "BCB420 Assignment 2"
author: "Kevin Lu"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
bibliography: A2.bib
csl: biomed-central.csl
nocite: "@*"
---

I'm using the same dataset from A1, ["Differential effect of SARS-CoV-2 spike glycoprotein 1 on human bronchial and alveolar lung mucosa models: Implications on pathogenicity."](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE185657),
published December 13, 2021. The corresponding publication can be found on GEO and in the [references](#references).

This experiment compares the expression response of human alveolar and bronchial
cells to SARS-CoV-2 spike protein through RNA sequencing. Figure 1 from the paper
shows the experimental design.

![Figure 1. Experiment design. Rahman et al, 2021.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8708014/bin/viruses-13-02537-g001.jpg)

```{r, include=FALSE}
# From A1
library(GEOmetadb)
library(GEOquery)
# Download the metadata GEO series, or load from disk if already downloaded
if (!file.exists("GSE185657.rds")) {
  gse <- getGEO("GSE185657")
  saveRDS(gse, "GSE185657.rds")
} else {
  gse <- readRDS("GSE185657.rds") 
}
# Download supplementary tarball
if (!file.exists("GSE185657_RAW.tar")) {
  files <- getGEOSuppFiles("GSE185657", makeDirectory=FALSE)
  target <- rownames(files[1])
} else{
  target <- "./GSE185657_RAW.tar"
}
contents <- untar(target, list=TRUE)
if (!file.exists("GSE185657_normalized_CPM.rds")) {
  # Remove all extracted files to reperform the extraction and decompression
  if (!file.exists("GSM5621226_A_C1.txt")) {
    untar(target)
    lapply(contents, gunzip)
  }
  # Construct the raw counts matrix
  targets <- gsub(".gz$", "", contents)
  raw_list <- lapply(targets, read.delim)
  genes <- raw_list[[1]]$Ensembl_gene_ID
  samples <- sapply(raw_list, function (sample) colnames(sample)[2])
  counts <- sapply(raw_list, function(sample) sample[[2]])
  rownames(counts) <- genes
  colnames(counts) <- samples
  saveRDS(counts, "GSE185657_raw_CPM.rds")
  # Map to HUGO gene symbols
  GRCh38 <- biomaRt::useEnsembl("genes", "hsapiens_gene_ensembl")
  ensg2hgnc <- biomaRt::getBM(c("ensembl_gene_id", "hgnc_symbol"), mart=GRCh38)
  gene_symbols <- ensg2hgnc$hgnc_symbol[match(genes, ensg2hgnc$ensembl_gene_id)]
  saveRDS(gene_symbols, "GSE185657_raw_symbols.rds")
  # Clean data based on read threshold
  filtered_counts <- counts[rowSums(counts > 1) >= 6, ]
  filtered_symbols <- ensg2hgnc$hgnc_symbol[match(rownames(filtered_counts), ensg2hgnc$ensembl_gene_id)]
  saveRDS(filtered_counts, "GSE185657_filtered_CPM.rds")
  saveRDS(filtered_symbols, "GSE185657_filtered_symbols.rds")
  # Apply TMM normalization to cleaned CPMs
  groups <- gsub("\\d", "", colnames(filtered_counts))
  d <- edgeR::DGEList(filtered_counts, group = groups)
  d <- edgeR::calcNormFactors(d)
  normalized_counts <- edgeR::cpm(d)
  saveRDS(normalized_counts, "GSE185657_normalized_CPM.rds")
} else {
  counts <- readRDS("GSE185657_raw_CPM.rds")
  gene_symbols <- readRDS("GSE185657_raw_symbols.rds")
  filtered_counts <- readRDS("GSE185657_filtered_CPM.rds")
  filtered_symbols <- readRDS("GSE185657_filtered_symbols.rds")
  normalized_counts <- readRDS("GSE185657_normalized_CPM.rds")
}
```

## Differential gene expression

Recall the design of this experiment. There are 24 samples. **A** samples are
alveolar cells and **B** samples are bronchial cells. **\_Sp** samples were exposed
a SARS-CoV-2 spike protein and **\_C** samples were controls.

```{r}
knitr::kable(t(normalized_counts[1:3,]), type="html")
```

Construct the design matrix according to the experiment design for differential
expression analysis.

```{r}
design <- data.frame(lapply(colnames(normalized_counts), function(x) {
  gsub("\\d", "", unlist(strsplit(x, "_")))
}))
colnames(design) <- colnames(normalized_counts)
rownames(design) <- c("cell_type", "exposure")
design <- data.frame(t(design))
knitr::kable(design, type="html")
```

Now we can answer the first two questions for this section:

1. Calculate p-values for each of the genes in your expression set. How many genes were significantly differentially expressed? What thresholds did you use and why?
2. Multiple hypothesis testing - correct your p-values using a multiple hypothesis correction method. Which method did you use? And Why? How many genes passed correction?

I'll be using the _limma_ package, since_lmFit_ directly supports accounting for
multiple confounders, which we have in this dataset. Following the approach in
lecture, using Benjamini-Hochberg correction for multiple hypothesis testing since
it is less stringent:

```{r}
design_model <- model.matrix(~ design$cell_type + design$exposure)
minimal_set <- ExpressionSet(normalized_counts)
fit <- limma::lmFit(minimal_set, design_model)
fit2 <- limma::eBayes(fit, trend=TRUE)
topfit <- limma::topTable(fit2, coef=ncol(design_model), adjust.method="BH", number=nrow(normalized_counts))
# Add the gene symbols
output_hits <- merge(
  data.frame(gene = filtered_symbols, row.names = rownames(normalized_counts)),
  topfit,
  by=0,
  all=TRUE
)
# Sort by unadjusted p-value
output_hits <- output_hits[order(output_hits$P.Value),]
knitr::kable(output_hits[1:10,], type="html", row.names=FALSE)
```

Choose the standard significance threshold 0.05. How many genes were significantly differentially expressed? 
```{r}
sum(output_hits$P.Value < 0.05)
```

How many genes passed correction?
```{r}
sum(output_hits$adj.P.Val < 0.05)
```

3. Show the amount of differentially expressed genes using an MA Plot or a Volcano plot. Highlight genes of interest.

```{r}
limma::plotMA(minimal_set, status=output_hits$P.Value < 0.05, main="Genes with differential expression p-value < 0.05")
```

4. Visualize your top hits using a heatmap. Do you conditions cluster together? Explain why or why not.

```{r}
top_hits <- output_hits$Row.names[output_hits$P.Value < 0.05]
heatmap_matrix_tophits <- t(scale(t(normalized_counts[which(rownames(normalized_counts) %in% top_hits),])))
heatmap_col <- circlize::colorRamp2(c(min(heatmap_matrix_tophits), 0, max(heatmap_matrix_tophits)), c("blue", "white", "red"))
heatmap <- ComplexHeatmap::Heatmap(as.matrix(heatmap_matrix_tophits),
  cluster_rows = TRUE, show_row_dend = TRUE,
  cluster_columns = FALSE,show_column_dend = FALSE,
  col=heatmap_col,show_column_names = TRUE,
  show_row_names = FALSE,show_heatmap_legend = TRUE)
heatmap
```

The conditions are definitely clustering together. In this heatmap, samples of the
same cell type and exposure are adjacent and the boundary between groups is
visually discernible.

## Thresholded over-representation analysis

1. Which method did you choose and why?

I'm choosing [g:Profiler](https://biit.cs.ut.ee/gprofiler/gost) due to familiarity and convenience.

2. What annotation data did you use and why? What version of the annotation are you using?

I can use the same sets of pathways as the homework as they are very relevant to
the behaviour being studied, which is the cellular response to the COVID spike protein.
I'll be using the latest version available from the g:Profiler site.

- GO biological process (GO:BP)
- Reactome (REAC)
- WikiPathways (WP)

3. How many genesets were returned with what thresholds?

Using the same p-values as before...

Up-regulated:
```{r}
sum(output_hits$P.Value < 0.05 & output_hits$logFC > 0)
```

Down-regulated
```{r}
sum(output_hits$P.Value < 0.05 & output_hits$logFC < 0)
```

```{r}
upreg_ensg <- output_hits$Row.names[output_hits$P.Value < 0.05 & output_hits$logFC > 0]
upreg_hgnc <- output_hits$gene[output_hits$P.Value < 0.05 & output_hits$logFC > 0]
downreg_ensg <- output_hits$Row.names[output_hits$P.Value < 0.05 & output_hits$logFC < 0]
downreg_hgnc <- output_hits$gene[output_hits$P.Value < 0.05 & output_hits$logFC < 0]
write.table(upreg_ensg, file = "upreg_ensg.txt", sep="\n", row.names = FALSE,col.names = FALSE,quote = FALSE)
write.table(upreg_hgnc, file = "upreg_hgnc.txt", sep="\n", row.names = FALSE,col.names = FALSE,quote = FALSE)
write.table(downreg_ensg, file = "downreg_ensg.txt", sep="\n", row.names = FALSE,col.names = FALSE,quote = FALSE)
write.table(downreg_hgnc, file = "downreg_hgnc.txt", sep="\n", row.names = FALSE,col.names = FALSE,quote = FALSE)
```

4. Run the analysis using the up-regulated set of genes, and the down-regulated set of genes separately. How do these results compare to using the whole list (i.e all differentially expressed genes together vs. the up-regulated and down regulated differentially expressed genes separately)?

Following the same steps as the homework, choosing the aforementioned data sources
and choosing "Benjamini-Hochberg FDR" for consistency with the prior correction method,
here are the top results of the query.

Up-regulated only:
- GO:BP: defense response to virus, defense response to symbiont, response to virus
- REAC: Interferon alpha/beta signaling
- WP: Type II interferon signaling (IFNG)

Down-regulated only:
- GO-BP: cytoplasmic translation
- REAC: Eukaryotic Translation Elongation
- WP: Cytoplasmic ribosomal proteins

Both:
- GO-BP: cellular macromolecule metabolic process
- REAC: Eukaryotic Translation Elongation
- WP: Cytoplasmic ribosomal proteins

Since the down-regulated list is larger, I think its influence dominated the combined query.

## Interpretation

1. Do the over-representation results support conclusions or mechanism discussed in the original paper?

Yes, the authors say that they "observed a typical anti-viral response in the bronchial model whereas a pro-fibrotic response in the alveolar model", which is consistent with the identified pathways.

2. Can you find evidence, i.e. publications, to support some of the results that you see. How does this evidence support your results.

Not answered due to time.

## References

Rahman M, Irmler M, Keshavan S, Introna M et al. Differential Effect of SARS-CoV-2 Spike Glycoprotein 1 on Human Bronchial and Alveolar Lung Mucosa Models: Implications for Pathogenicity. Viruses 2021 Dec 17;13(12). PMID: [34960806](https://pubmed.ncbi.nlm.nih.gov/34960806/)

Uku Raudvere, Liis Kolberg, Ivan Kuzmin, Tambet Arak, Priit Adler, Hedi Peterson, Jaak Vilo: g:Profiler: a web server for functional enrichment analysis and conversions of gene lists (2019 update) Nucleic Acids Research 2019; doi:10.1093/nar/gkz369

<div id="refs"></div>

## Journal

[GitHub wiki](https://github.com/bcb420-2022/Kevin_Lu/wiki/Assignment-2)
