---
title: "Differential gene expression and preliminary ORA on GSE185657"
subtitle: "BCB420 Assignment 2"
author: "Kevin Lu"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
bibliography: A2.bib
csl: biomed-central.csl
nocite: "@*"
---

```{r, include=FALSE}
# From A1
library(GEOmetadb)
library(GEOquery)
# Download the metadata GEO series, or load from disk if already downloaded
if (!file.exists("GSE185657.rds")) {
  gse <- getGEO("GSE185657")
  saveRDS(gse, "GSE185657.rds")
} else {
  gse <- readRDS("GSE185657.rds") 
}
# Download supplementary tarball
if (!file.exists("GSE185657_RAW.tar")) {
  files <- getGEOSuppFiles("GSE185657", makeDirectory=FALSE)
  target <- rownames(files[1])
} else{
  target <- "./GSE185657_RAW.tar"
}
contents <- untar(target, list=TRUE)
if (!file.exists("GSE185657_normalized_CPM.rds")) {
  # Remove all extracted files to reperform the extraction and decompression
  if (!file.exists("GSM5621226_A_C1.txt")) {
    untar(target)
    lapply(contents, gunzip)
  }
  # Construct the raw counts matrix
  targets <- gsub(".gz$", "", contents)
  raw_list <- lapply(targets, read.delim)
  genes <- raw_list[[1]]$Ensembl_gene_ID
  samples <- sapply(raw_list, function (sample) colnames(sample)[2])
  counts <- sapply(raw_list, function(sample) sample[[2]])
  rownames(counts) <- genes
  colnames(counts) <- samples
  saveRDS(counts, "GSE185657_raw_CPM.rds")
  # Map to HUGO gene symbols
  GRCh38 <- biomaRt::useEnsembl("genes", "hsapiens_gene_ensembl")
  ensg2hgnc <- biomaRt::getBM(c("ensembl_gene_id", "hgnc_symbol"), mart=GRCh38)
  gene_symbols <- ensg2hgnc$hgnc_symbol[match(genes, ensg2hgnc$ensembl_gene_id)]
  saveRDS(gene_symbols, "GSE185657_raw_symbols.rds")
  # Clean data based on read threshold
  filtered_counts <- counts[rowSums(counts > 1) >= 6, ]
  filtered_symbols <- ensg2hgnc$hgnc_symbol[match(rownames(filtered_counts), ensg2hgnc$ensembl_gene_id)]
  saveRDS(filtered_counts, "GSE185657_filtered_CPM.rds")
  saveRDS(filtered_symbols, "GSE185657_filtered_symbols.rds")
  # Apply TMM normalization to cleaned CPMs
  groups <- gsub("\\d", "", colnames(filtered_counts))
  d <- edgeR::DGEList(filtered_counts, group = groups)
  d <- edgeR::calcNormFactors(d)
  normalized_counts <- edgeR::cpm(d)
  saveRDS(normalized_counts, "GSE185657_normalized_CPM.rds")
} else {
  counts <- readRDS("GSE185657_raw_CPM.rds")
  gene_symbols <- readRDS("GSE185657_raw_symbols.rds")
  filtered_counts <- readRDS("GSE185657_filtered_CPM.rds")
  filtered_symbols <- readRDS("GSE185657_filtered_symbols.rds")
  normalized_counts <- readRDS("GSE185657_normalized_CPM.rds")
}
```


## Differential gene expression

1. Calculate p-values for each of the genes in your expression set. How many genes were significantly differentially expressed? What thresholds did you use and why?
1. Multiple hypothesis testing - correct your p-values using a multiple hypothesis correction method. Which method did you use? And Why? How many genes passed correction?
1. Show the amount of differentially expressed genes using an MA Plot or a Volcano plot. Highlight genes of interest.
1. Visualize your top hits using a heatmap. Do you conditions cluster together? Explain why or why not.

## Thresholded over-representation analysis

1. Which method did you choose and why?
1. What annotation data did you use and why? What version of the annotation are you using?
1. How many genesets were returned with what thresholds?
1. Run the analysis using the up-regulated set of genes, and the down-regulated set of genes separately. How do these results compare to using the whole list (i.e all differentially expressed genes together vs. the up-regulated and down regulated differentially expressed genes separately)?

## Interpretation

1. Do the over-representation results support conclusions or mechanism discussed in the original paper?
1. Can you find evidence, i.e. publications, to support some of the results that you see. How does this evidence support your results.

## References

Rahman M, Irmler M, Keshavan S, Introna M et al. Differential Effect of SARS-CoV-2 Spike Glycoprotein 1 on Human Bronchial and Alveolar Lung Mucosa Models: Implications for Pathogenicity. Viruses 2021 Dec 17;13(12). PMID: [34960806](https://pubmed.ncbi.nlm.nih.gov/34960806/)

<div id="refs"></div>

## Journal

[GitHub wiki](https://github.com/bcb420-2022/Kevin_Lu/wiki/Assignment-2)
